File: languages/c/solve_1.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

void bubble_sort(unsigned int (*array)[1024], int length) {
  bool sorted = false;
  while (!sorted) {
    bool in_order = true;
    for (int i = 0; i < length - 1; i++) {
      int val = (*array)[i];
      int next = (*array)[i + 1];
      if (val > next) {
        in_order = false;
        (*array)[i] = next;
        (*array)[i + 1] = val;
      }
    }
    if (in_order) {
      sorted = true;
    }
  }
}



int main() {
  FILE *file =
      fopen("/home/marcus/Documents/coding/advent-of-code/2024/inputs/1.txt",
            "r"); // Open the file in read mode
  if (file == NULL) {
    perror("Error opening file");
    return 1; // Exit if the file cannot be opened
  }
  char *line = NULL;
  size_t len = 0;
  unsigned int array_index = 0;

  unsigned int list_one[1024];
  unsigned int list_two[1024];

  // Read the file line by line
  while (getline(&line, &len, file) != -1) {
    char num_string_one[6];
    char num_string_two[6];
    for (int i = 0; i < 5; i++) {
      num_string_one[i] = line[i];
      num_string_two[i] = line[i + 8];
    }
    num_string_one[5] = '\0';
    num_string_two[5] = '\0';
    list_one[array_index] = atoi(num_string_one);
    list_two[array_index] = atoi(num_string_two);
    array_index += 1;
  }

  fclose(file); // Close the file when done
  free(line);

  bubble_sort(&list_one, array_index);
  bubble_sort(&list_two, array_index);

  int diff = 0;
  for (int i = 0; i < array_index; i++) {
    int entry_list_one = list_one[i];
    int entry_list_two = list_two[i];
    diff += abs(entry_list_one - entry_list_two);
  }

  printf("Part One: %d\n", diff);

  int frequency_arr[100000] = {0};
  for (int i = 0; i < array_index; i++) {
      frequency_arr[list_two[i]]++;
  }

  int similarity = 0;

  for (int i = 0; i < array_index; i++) {
      int val = list_one[i];
      similarity += val * frequency_arr[val];
  }

  printf("Part Two: %d\n", similarity);
  return 0;
}

File: languages/javascript/solve_1.mjs
import { readFileSync } from "fs";

const file = readFileSync(
  "/home/marcus/Documents/coding/advent-of-code/2024/inputs/1.txt",
  { encoding: "utf-8" },
);

function parseInput(input) {
  return input
    .split("\n")
    .filter(Boolean)
    .map((vals) => vals.split("  ").filter(Boolean))
    .map((vals) => vals.map((val) => parseInt(val, 10)))
    .reduce(
      (acc, [one, two]) => {
        acc[0].push(one);
        acc[1].push(two);
        return acc;
      },
      [[], []],
    );
}

function partOne(listOne, listTwo) {
  [listOne, listTwo].map((list) => {
    list.sort((a, b) => a - b);
    return list;
  });
  let difference = 0;
  listOne.forEach((listOneEntry, index) => {
    const listTwoEntry = listTwo[index];
    difference += Math.abs(listOneEntry - listTwoEntry);
  });
    return difference;
}

/**
    * @param {number[]} listOne 
    * @param {number[]} listTwo 
    */
function partTwo(listOne, listTwo) {
    const countMap = new Map();
    listTwo.forEach((num)=>{
        const current = countMap.get(num) ?? 0;
        countMap.set(num, current+1);
    })
    return listOne.reduce((acc,num)=>acc + num * (countMap.get(num) ?? 0),0);
}

const [listOne, listTwo] = parseInput(file);
const answerOne = partOne(listOne, listTwo);
const answerTwo = partTwo(listOne,listTwo);
console.log({answerOne, answerTwo})
